include/parallel/parallel_implementation.h

```diff
@@ -2966,9 +2967,13 @@ inline void Communicator::gather(const unsigned int root_id,
         sendlengths  (this->size(), 0),
         displacements(this->size(), 0);
 
-      // first comm step to determine buffer sizes from all processors
       const int mysize = static_cast<int>(sendval.size());
-      this->gather(root_id, mysize, sendlengths);
+
+      if (identical_buffer_sizes)
+        sendlengths.assign(this->size(), mysize);
+      else
+        // first comm step to determine buffer sizes from all processors
+        this->gather(root_id, mysize, sendlengths);
 
```

Comment:
We could also use MPI_Gather instead of MPI_Gatherv in the identical_buffer_sizes case, but I suspect just avoiding the latency of the preliminary gather is 95% of the possible optimization here.